From 9ccbbb18cfcf3ab5cee652a826f2c47036b9b4c4 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:11:19 -0400
Subject: [powerpc] powernv: Check sysfs size before copying

Message-id: <1410545655-205645-451-git-send-email-dzickus@redhat.com>
Patchwork-id: 94353
O-Subject: [RHEL7 PATCH 450/626] powerpc/powernv: Check sysfs size before copying
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 85390378f0a6abf3bdb10c3d504d77862a2d0ec9
Author: Joel Stanley <joel@jms.id.au>
Date:   Thu Apr 24 16:55:35 2014 +0930

    powerpc/powernv: Check sysfs size before copying

    The sysparam code currently uses the userspace supplied number of
    bytes when memcpy()ing in to a local 64-byte buffer.

    Limit the maximum number of bytes by the size of the buffer.

    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/powernv/opal-sysparam.c b/arch/powerpc/platforms/powernv/opal-sysparam.c
index 0675e1a..83f5f2a 100644
--- a/arch/powerpc/platforms/powernv/opal-sysparam.c
+++ b/arch/powerpc/platforms/powernv/opal-sysparam.c
@@ -135,6 +135,10 @@ static ssize_t sys_param_store(struct kobject *kobj,
    kobj_attr);
  ssize_t ret;
 
+        /* MAX_PARAM_DATA_LEN is sizeof(param_data_buf) */
+        if (count > MAX_PARAM_DATA_LEN)
+                count = MAX_PARAM_DATA_LEN;
+
  mutex_lock(&opal_sysparam_mutex);
  memcpy(param_data_buf, buf, count);
  ret = opal_set_sys_param(attr->param_id, attr->param_size,
-- 
1.7.1